libcudacxx/include/cuda/std/__random/philox_engine.h

```diff
@@ -0,0 +1,575 @@
+  //! @return is
+  template <typename _CharT, typename _Traits>
+  _CCCL_API friend ::std::basic_istream<_CharT, _Traits>&
+  operator>>(::std::basic_istream<_CharT, _Traits>& __is, philox_engine& __e)
+  {
+    using istream_type = ::std::basic_istream<_CharT, _Traits>;
+    using ios_base     = typename istream_type::ios_base;
+
+    // save old flags
+    const typename ios_base::fmtflags __flags = __is.flags();
+
+    __is.flags(ios_base::dec);
+
+    // input counter array (__x_)
+    for (::cuda::std::size_t __i = 0; __i < word_count; ++__i)
+    {
+      __is >> __e.__x_[__i];
+    }
+
+    // input key array (__k_)
+    for (::cuda::std::size_t __i = 0; __i < word_count / 2; ++__i)
+    {
+      __is >> __e.__k_[__i];
+    }
+
+    // input output buffer (__y_)
+    for (::cuda::std::size_t __i = 0; __i < word_count; ++__i)
+    {
+      __is >> __e.__y_[__i];
+    }
+
+    // input current position
+    __is >> __e.__j_;
+
+    // restore flags
+    __is.flags(__flags);
+
+    return __is;
+  }
+#endif // !_CCCL_COMPILER(NVRTC)
+
+private:
+  _CCCL_API constexpr void __increment_counter() noexcept
+  {
+    // Increment the big integer __x_ by 1, handling overflow.
+    _CCCL_PRAGMA_UNROLL_FULL()
+    for (::cuda::std::size_t __i = 0; __i < word_count; ++__i)
+    {
+      __x_[__i] = (__x_[__i] + 1) & max();
+      if (__x_[__i] != 0)
+      {
+        break;
+      }
+    }
+  }
+
+  [[nodiscard]] _CCCL_API constexpr auto __mulhilo_fallback(result_type __a, result_type __b) const noexcept
+  {
+    // Generic slow implementation
+    constexpr result_type __w_half  = word_size / 2;
+    constexpr result_type __lo_mask = (((result_type) 1) << __w_half) - 1;
+
+    result_type __lo  = __a * __b;
+    result_type __ahi = __a >> __w_half;
+    result_type __alo = __a & __lo_mask;
+    result_type __bhi = __b >> __w_half;
+    result_type __blo = __b & __lo_mask;
+
+    result_type __ahbl = __ahi * __blo;
+    result_type __albh = __alo * __bhi;
+
+    result_type __ahbl_albh = ((__ahbl & __lo_mask) + (__albh & __lo_mask));
+    result_type __hi        = __ahi * __bhi + (__ahbl >> __w_half) + (__albh >> __w_half);
+    __hi += __ahbl_albh >> __w_half;
+    __hi += ((__lo >> __w_half) < (__ahbl_albh & __lo_mask));
+
+    return ::cuda::std::make_pair(__hi, __lo);
+  }
+
+  [[nodiscard]] _CCCL_API constexpr auto __mulhilo(result_type __a, result_type __b) const noexcept
+  {
+    if constexpr (word_size == 32)
+    {
+      // std::uint_fast32_t can actually be 64 bits so cast to 32 bits
+      return ::cuda::std::make_pair(
+        static_cast<result_type>(::cuda::mul_hi(static_cast<std::uint32_t>(__a), static_cast<std::uint32_t>(__b))),
+        (__a * __b) & max());
+    }
+    else if constexpr (word_size == 64)
+    {
+      auto __hi =
+        static_cast<result_type>(::cuda::mul_hi(static_cast<std::uint64_t>(__a), static_cast<std::uint64_t>(__b)));
+      auto __lo = (__a * __b) & max();
```

Comment:
I believe we should split the first clause the same as the second one to make it easier to read. 

Also just use `::cuda::std::pair`